### PR TITLE
For #391: created conditional effect

### LINKED ISSUE #391: Implement Conditional Effects with Dynamic Conditions
The puzzle `361-056160f5` from EnigmaDragons/ProjectNeon#361 has to be resolved: 

https://github.com/EnigmaDragons/ProjectNeon/blob/01b1db7cee00c260127b1ea50b079598f8c1bf96/src/ProjectNeon/Assets/Features/Effects/AllEffects.cs#L28-L28

The puzzle was created by Paulo Lobo on 23-Oct-19. 

Estimate: 30 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by paulodamaso**
Closes #391: Creation of conditional effects
- Created conditional effects decorator
- Created tests

### CODE REVIEW

**Comment by SilasReinagel** [comment]
@paulodamaso Is this PR still relevant?

If something new is being coded, let's close this PR and open a new one later.

**Comment by paulodamaso** [comment]
@SilasReinagel I'll maybe need architectural changes for this one, but I'm working on it

**Comment by paulodamaso** [comment]
@SilasReinagel Implemented correctly the card:
- Changed the image to a more adequate one
- Implemented `ConditionalEffect` , which will allow us to compose cards with conditional execution
- Implemented `EffectIfAnyHealthBelowThreshold` state, which is a conditional effect that only happens when any effect target has health below a % threshold

We have reached the architecture limit for effect composition: we can't compose effects which have more parameters than the described in `EffectData` That leaves us to a maximum of two parameters, one float and one int (`NumberOfTurns`). So we can't compose two effects if both of them use the same parameter. Note that I had to use `NumberOfTurns` parameter to pass the shield quantity to be added; if this is bad, the only other way around is to make this value fixed for this conditioinal effect.

**Comment by paulodamaso** [comment]
@SilasReinagel I've got an idea for solving some of these problems, please do not merge yet

**Comment by SilasReinagel** [comment]
@paulodamaso Passing arbitrary args will make this impossible to use for our Designers. 

Designers should not be required to know and memorize code in order to create their cards.

Also, the current direction of this PR doesn't seem related to the original ticket.

**Comment by paulodamaso** [comment]
@SilasReinagel The original ticket needs a update on Effect logic, so we can pass more parameters and / or compose the effects better, so my intent is to solve update the logic and test it solvin #391

**Comment by paulodamaso** [comment]
@SilasReinagel Cleaned the PR and made the implementation based on #467

File: src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+﻿
+using System;
+using UnityEngine;
+
+public abstract class ConditionalEffect : Effect
+{
+    protected Effect _effect;
+    protected Member _source;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/ConditionalEffect.cs]: This same comment is on all the PRs.

Don't store `Member` and `Target`.

File: src/ProjectNeon/Assets/Features/Effects/AllEffects.cs [hunk 2]
```diff
@@ -35,9 +35,11 @@ public static class AllEffects
         { EffectType.BuffToughnessFlat, e => new SimpleEffect(m => m.ApplyTemporaryAdditive(new BuffedStats(new StatAddends().With(StatType.Toughness, e.IntAmount), e.NumberOfTurns)))},
         { EffectType.HealFlatForTurnsOnTurnStart, e => new HealFlatForTurnsOnTurnStart(e.IntAmount, e.NumberOfTurns) },
         { EffectType.RepeatEffect, e => new RepeatEffect(Create(e.origin), e.IntAmount) },
-        { EffectType.RandomizeTarget, e => new RandomizeTarget(Create(e.origin)) }
+        { EffectType.RandomizeTarget, e => new RandomizeTarget(Create(e.origin)) },
+        { EffectType.AnyTargetHealthBelowThreshold, e => new AnyTargetHealthBelowThreshold(Create(e.origin), e.FloatAmount) }
 
-        //deal .5 attack for a random enemy 3 times
+
+        //If they are still below 50% life also Shield 5.
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/AllEffects.cs]: Dead comment?

File: src/ProjectNeon/Assets/Features/Effects/AnyTargetHealthBelowThreshold.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    private float _threshold;
+
+    public AnyTargetHealthBelowThreshold (Effect effect, float threshold) : base(effect)
+    {
+        _threshold = threshold;
+    }
+
+    public override bool Condition()
+    {
+        return Array.Exists(_target.Members, element => element.State[TemporalStatType.HP] < element.State.MaxHP() * _threshold );
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Features/Effects/AnyTargetHealthBelowThreshold.cs]: It's generally better to use `System.Linq`'s `.Any()` extension method for this. 

Also, for short lambdas, it's nicer to use a single-letter variable name such as `x` or `e`.

**Pull request merged**